### PR TITLE
Add I/O interfaces to Roadmap.md

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

File: Roadmap.md [hunk 1]
```diff
@@ -243,3 +243,118 @@ Fetch API:
 ```ts
 fetch(input?: Request | string, init?: RequestInit): Promise<Response>;
 ```
+
+#### I/O
+
+There will be two API layers for I/O:
+1. A low-level abstraction with (for sockets) non-blocking read/writes operating
```
Comment (gillchristian) [inline Roadmap.md]: > A low-level abstraction ~with~ (for sockets) with non-blocking read/writes

File: Roadmap.md [hunk 2]
```diff
@@ -243,3 +243,162 @@ Fetch API:
...
+}
+
+// Low-level non-blocking, non-async read.
+function read(fd: number, p: Uint8Array, nbytes: number): ReadResult;
+
+// Low-level non-blocking, non-async write.
+function write(fd: number, p: Uint8Array, nbytes: number): number;
+
+// Low-level close.
+function close(fd: number): void;
```
Comment (qti3e) [inline Roadmap.md]: I think you've missed `open`
```ts
open(pathname: string, flags: OpenFlags): number;
```

**Comment by benjamingr** [comment]
Hey, - I'm not sure this is the correct approach for the high level API because there are very nice native constructs we can utilize instead (the low level API looks good).

Namely, instead of defining a new `Reader` interface with a `read`  returning a promise. An `AsyncIterator` can be used. The DOM has a BYOB (bring your own buffer) reader as well that does/will support async iterators.

Node also does this with its streams in Node.js 10. Basically instead of doing the following which the proposed solution does:

```ts
async function copy(dst: Writer, src: Reader): Promise<number> {
    let n = 0;
    const b = new Uint8Array(1024);
    let got_eof = false;
    while (got_eof === false) {
      let result = await src.read(b);
      if (result.eof) got_eof = true;
      n += await dst.write(b.subarray(0, result.nread));
   }
  return n;
}
```

You could do:
```ts
async function copy(dst: Writer, src: Reader): Promise<number> {
  let n = 0;
  for await(const chunk of reader) { // chunks reuse the same memory internally
    await dst.write(chunk);
    n += dst.byteLength;
    // doing `break;` here calls `.return` on the iterator automatically
  }
  return n; // not really needed since the state is on the iterator now and not a number which is safer.
}
```

This is especially convenient for two reasons:

 - It integrates seamlessly with async/await and tooling and is part of the language itself. 
 - It works beautifully with workflows for sockets.
 - It's easier to maintain since it's a part of the language and lives in v8 mostly.

**Comment by ry** [comment]
@benjamingr I don't see the benefit of your proposal. Your solution is a higher level, because the buffer allocation is now part of the interface. With the API I'm proposing, allocation is handled by the user -- which is useful in many applications where the user has specific knowledge of how the memory will be used.

**Comment by benjamingr** [comment]
I totally see the value of the user controlling allocation and in fact I think any good I/O implementation should provide that capability.  This is what I meant with [BYOB reader](https://streams.spec.whatwg.org/#byob-readers)). The above example assumed that the buffer was passed when the reader was acquired.

It is also possible to do by passing it to the part acquiring the iteration [like this](https://gist.github.com/benjamingr/b08c4e9b43ac83313d6cb52482cbdf14).

The rationale for using the built in standard JavaScript mechanism for this (async iterators) is that the language already allows for a construct where:
 - Iteration is done asynchronously and returns promises.
 - Model is Pull based and backpressure works out of the box.
 - Closing and exception handling work out of the box.
 - Tooling for debugging issues are easier to use and rely on the language itself.

I definitely think that any reasonable API should let the user control allocation as this is something that personally pained me a lot of times.

**Comment by ry** [comment]
@benjamingr 

> This is what I meant with BYOB reader.

Oh, I missed that. Nevertheless this all seems way more complex than what I've outlined.

> Iteration is done asynchronously and returns promises.
> Model is Pull based and backpressure works out of the box.
> Closing and exception handling work out of the box.
> Tooling for debugging issues are easier to use and rely on the language itself.

None of these points are excluded in my proposal.

Since my proposal is lower-level and simpler. I think it would make more sense for users to implement the streams standard on top of this.

**Comment by ry** [comment]
@benjamingr Maybe "lower-level" isn't the right term since it seems you could implement this Go interface in terms of "Standard Streams". It's simpler. 

The ideal of UNIX is this concept where "everything is a file". That description isn't quite right... The ideal of UNIX is that FDs are these core abstractions that can be manipulated and combined for all sorts of purposes. The ideal FD interface is just read/write/open/close. We never quite got this ideal. Linux is massively more complex than that - but everyone considers it a laudable goal. The Go I/O interfaces are basically the simplest lowest common denominator that supports all real world situations: async File I/O, non-blocking sockets, half-opened connections, backpressure. This is the interface Rob Pike returned with 20 years after Plan9. 

The "Standard Streams" proposal is unnecessarily complex.

I have nothing against async iterators - but just because they are possible doesn't mean that everything should be an async iterator. "read" is a fundamental operation and should be exposed as such.

**Comment by ry** [comment]
The main question to me is how do we adapt these Go interfaces to JavaScript. For example: what errors are treated as exceptions, if any?

**Comment by benjamingr** [comment]
@ry to be clear: I'm not saying deno should implement the whatwg standards proposal.

I also totally get why the Go interfaces make sense _for Go_. I also think they would make an OK interface _for TypeScript_ with the right semantics. I also like the unix "everything is a file" philosophy - as you said not everything is random access, readable and writeable. Let's ignore for a moment anything but non-random-access readable things.

For _that case_ JavaScript already ships with async iterators which have very similar semantics to what you are describing (for reading). 

The interface is basically having a `.next` method (which may take a parameter in if you choose):
```js
const reader = getReader(); // get async iterator from somewhere
await reader.next(buffer); // returns Promise<{done: false }>, can also return other data
reader.return(); // this closes the underlying reader
reader.throw(); // may also choose to support it
```

I have some slides in this pretty old [2016 talk](https://docs.google.com/presentation/d/1r2V1sLG8JSSk8txiLh4wfTkom-BoOsk52FgPBy8o3RM/edit#slide=id.g1236b1e2d6_0_7) that predates it making the standard.

Basically all it requires is that `.read` on the reader is called `.next`, a buffer is optionally signed in and that it returns a Promise for an iteration result and not a number. BYOB can be built in (and the only choice). 

I see async iterators as a way to "JavaScript"ize Go's Reader. They are _not_ the only way but I think it's really nice to happen to have a built-in construct that happens to have very reasonable reader semantics.

**Comment by benjamingr** [comment]
> It will be very easy to wrap any "Reader" interface (as I've defined above) to do this.

I think we can agree it will be very easy to wrap every interface you (or I) come up with with probably every other interface that has reasonable pull properties.

>  Implementing "read" is a separate concern to whether you want to iterate on that object.

I am not sure I understand that fundamental difference between iteration and reading at all. My understanding is that iteration just encapsulates the state management in the iterator (managing the number of bytes read) whereas reading does not.

My concern here is that by not embracing the language's abstractions (like the iteration protocol) Deno would be working 'against the grain' as the language grows - much like how Node is stuck with callback APIs forever now while everyone is using async/await.

Again - I totally understand why the Go APIs make sense for Go.

Also - It's totally fine if you don't see it that way or think that the problem of iterating a file and the problem of reading it are sufficiently difference or if you think the API is footgunny or bad. I just feel like this might turn out to be a missed opportunity like promises in Node.


-----

By the way:  There is a nice paper by Erik Meijer that explains where all the concurrency primitives made sense when they did this with async/await (for dart?). I'll try to dig it up (hopefully the ping works). It's also worth trying to get Gilad Bracha to see if it's possible to get feedback about what modern readers/writers might look like with async/await. Of course - feedback from Rob Pike would also be nice given the vast amount of experience he has with this problem. 

Normally I'd just ping them (didn't want to be rude, is that ok with you?) - since this is a new platform the people who worked on the abstractions and APIs of other languages and platforms typically have a lot of interesting things to say that are worth exploring.

**Comment by ry** [comment]
> I am not sure I understand that fundamental difference between iteration and reading at all.

Iterating (in this context) is reading in a loop - read is a single read. Maybe you just want the first 10 bytes into a given buffer. Here this is 
```
let buf = new Uint8Array(10);
await reader.read(buf); 
```
which is pretty close to the natural language phrasing. How does this look in your proposal?

> My concern here is that by not embracing the language's abstractions (like the iteration protocol) Deno would be working 'against the grain' as the language grows. [...] I just feel like this might turn out to be a missed opportunity like promises in Node.

I acknowledge your concern. But I think async iterators is not the correct abstraction here. We're defining the foundational I/O interfaces, they should be as orthogonal as possible. 

That's not to say that I don't think async iterators shouldn't be used for I/O - I just don't think it should be forced at this layer. If someone has a class that they want to be used with the I/O utilities, like `copy`, should they need to implement an iterator? Or should they just implement `read`? 

> Normally I'd just ping them (didn't want to be rude, is that ok with you?)

That's fine with me. The point of this Roadmap is to get more eyes on it in order to make fewer design mistakes.

**Comment by benjamingr** [comment]
> That's fine with me. The point of this Roadmap is to get more eyes on it in order to make fewer design mistakes.


@headinthebox @robpike @gbracha - sorry for the ping but this has worked out too well in the past to not try. Deno is a new server-side platform written with TypeScript that is working on new I/O interfaces. We are currently discussing what those interfaces look like.

All proposed solutions are pull based (as in the consumer asks for the data in chunks), allow the consumer to control allocations and differ in what control flow structures they use. If you got to design these sort of APIs all over again - what would you do?

 - Gilad and Erik - thinking about what you did for Dart here and how that might look like with ECMAScript async iterators (or not if that's a poor choice).
 - Rob - the proposed design by Ryan is based off Go's Reader/Writer, does that make sense for JavaScript/TypeScript as a reasonable API to build on given Go's/TypeScript's differences?

If you prefer discussing this in email or in a smaller (or less public) forum - do feel free to reach out at benjamingr@gmail.com (or to Ryan who wrote the thing directly at ry@tinyclouds.com).

**Comment by benjamingr** [comment]
> Iterating (in this context) is reading in a loop - read is a single read. Maybe you just want the first 10 bytes into a given buffer. Here this is

While iterating in a loop is the most common case, just grabbing "the next item" is also a pretty common use case. I think it is important for any proposal to support both of these.

> which is pretty close to the natural language phrasing. How does this look in your proposal?

Well:

```js
let buf = new Uint8Array(10);
await fileIterator.next(buf); 
```

Which I concede is a bit uglier for this particular use case. It is also possible to expose `read` for reading things into a buffer as well as `next` for iterating the file (like python does) which would be the best of both worlds in terms of how the API looks but would be a bigger API surface.

>  If someone has a class that they want to be used with the I/O utilities, like copy, should they need to implement an iterator? Or should they just implement read?

Well, implementing `next` isn't more work than implementing read but I believe most people will implement neither and will use an async generator instead which does this automatically:

```js
// this example polls an 'async memory stream' that can be a file, database, or whatever
// it's only async since iteration itself is async and a synchronous memory stream would be pointless

// proposed API

function transformStream(stream, processFunction) {
  // omitting the need to check how many bytes we actually read here 
  // synchronise read calls since consumers are not guaranteed to await reads
  let queue = Promise.resolve(); 
  read(buffer) {
    queue = queue.then(async () => {
      await stream.read(buffer); // also need to check for eof here and propagate etc
      await processFunction(buffer);
    });
    await queue; 
  }
}

// async iterators
async function* transformStream(stream, processFunction) {
  // no need for queuing and synchronization
  // no need for eof check since closing the source will finish the loop
  // we can pass the buffer here if we want or omit it
  // and a reference to the underlying source will be used if possible with 0 copies
  for await(const chunk of stream) {
    yield await processFunction(chunk);
  }
}
```

**Comment by ry** [comment]
```
let buf = new Uint8Array(10);
await fileIterator.next(buf); 
```
And maybe only 5 bytes were actually read. How do I find that out? I suppose EOF is signaled by` { done: true }` ?

**Comment by benjamingr** [comment]
> And maybe only 5 bytes were actually read. How do I find that out?

The version of `next` that takes a buffer can return that through the value:

```js
let buf = new Uint8Array(10);
 // result.done for eof and result.value for the number of bytes read
 // you can also do result.n or result.nread if you prefer
let result = await fileIterator.next(buf);
```

> I suppose EOF is signaled by{ done: true } ?

Yes, and in general any reader finishing as well as any iterator returned from ES2018 code using `async function*` (or libraries like [axax](https://github.com/jamiemccrindle/axax)).

**Comment by ry** [comment]
@benjamingr Ok, so it's very close to the same interface - the function is called 'next' and returns `{ value, done }` instead of `read()` which returns `{ nread, eof }`. I still think it's better to call things as they are and let people wrap them in iterator if needed.

**Comment by benjamingr** [comment]
@ry thinking about this more - I think Reader should just expose a `Symbol.asyncIterator` in addition to `read`. I don't mind contributing the code for that.

That way people will be able to use both `for await(` and we get Reader-read instead of Iterator-next for cases like reading the first 10 bytes. That way people can still utilize libraries and the ecosystem of tools as well as the native platform tooling but the API would still be familiar to them coming from Go.

This is what the other two languages with async iterators already landed (Python and Dart) do as well.

Since it's only exposing a symbol (much like arrays expose Symbol.species and Symbol.iterator) it doesn't really increase the API surface people use or expect either.

File: Roadmap.md [hunk 3]
```diff
@@ -243,3 +243,151 @@ Fetch API:
...
+  // Callers should always process the n > 0 bytes returned before considering
+  // the error err. Doing so correctly handles I/O errors that happen after
+  // reading some bytes and also both of the allowed EOF behaviors.
+  //
+  // Implementations of Read are discouraged from returning a zero byte count
+  // with a nil error, except when p.byteLength == 0. Callers should treat a
+  // return of 0 and nil as indicating that nothing happened; in particular it
+  // does not indicate EOF.
+  //
+  // Implementations must not retain p.
```
Comment (piscisaureus) [inline Roadmap.md]: I'm not really sure what this means in this context.

File: Roadmap.md [hunk 4]
```diff
@@ -243,3 +243,151 @@ Fetch API:
...
+  async read(p: Uint8Array): Promise<ReadResult>;
```
Comment (piscisaureus) [inline Roadmap.md]: 1.  Maybe we should say `p: ArrayBufferView`, where 
    ```ts
    interface ArrayBufferView {
      buffer: ArrayBuffer,
      byteOffset: number,
      byteLength: number
    }
    ```

2.  I feel that `p` should be optional. Reading into a preallocated buffer is probably optimal if you can re-use the same buffer over and over again. But I think a lot of code will end up doing this:
    ```ts
    let buf = new Uint8Array(64436);
    const r = conn.read(buf);
    r = r.subarray(0, nread);
    ```

Comment (ry) [reply Roadmap.md]: +1 on ArrayBufferView

-1 ln optional arguments. This is meant to be the most minimal interface that objects must implement to be a Reader - this ergonomics are left to higher layers.

Comment (benjamingr) [reply Roadmap.md]: +1 on ArrayBufferView too.

If we expose `Symbol.asyncIterator` around `.read` that works with a view (and hopefully no buffer allocations at all but just a view into the memory) we get the minimal `.read` as well as the usable `for await`.

**Comment by ry** [comment]
@benjamingr I'd rather add utility function to convert any reader to an async iterator. Does this work for you?
```
import { readerIterator } from "deno";

for await (let buf of readerIterator(socket)) {
  console.log("read ${buf.byteLength} from socket");
}
```

File: Roadmap.md (hunk 3 above)
Comment (ry) [reply Roadmap.md]: It means that anyone who is implementing the Reader interface should not store copies of the TypedArrays passed to it.

**Comment by milutinovici** [comment]
I'd like to ping @davidfowl He designed the new IO primitives for .net core. 
It would be great if you could share your thoughts.

**Comment by robpike** [comment]
I don't know TypeScript well enough (at all, really), so have little to say besides showing the successful example of Go's I/O interfaces, which are already informing the conversation.

It is important to specify all the details around who owns the memory in the buffer. Also, define or explicitly leave undefined every possible detail about what happens on incomplete or failing or partially failing I/O operations.

**Comment by benjamingr** [comment]
@milutinovici oh sure, @davidfowl was a lot of help a few years ago talking interfaces (on jabbr?)!

@robpike thanks for weighing in! is there anything in particular you regret about Go's interfaces or you would do differently today?

@ry it's better than not having it at all since now users will be able to use `for await` without platform code. Would you mind explaining why it's better that way (as an external method and not a symbol)? 

The biggest pro I can think of for making it external is that it might make implementing readers easier. The con is that if readers are expected to be worked with directly then this will be very verbose whenever a stream is iterated (which I assume is a very common if not the most common case). It is also different from what other languages did.

Another concern is that if readers are not async-iterable people might pass around two types (readers and iterators). 

Other platforms solve that part with subclassing (which I don't really think we should do). It's also possible to expose a factory (like `makeReader` or something like that).

**Comment by davidfowl** [comment]
@milutinovici Thanks for pinging me. FWIW the new IO abstractions on .NET being referred to are [System.IO.Pipelines](https://blogs.msdn.microsoft.com/dotnet/2018/07/09/system-io-pipelines-high-performance-io-in-net/). They are a bit higher level but are trying to solve some of the hard problems you face when doing high performance IO.  

> It is important to specify all the details around who owns the memory in the buffer. 

+1, that's fundamental to making a good IO API. Most of the IO APIs in the various standard libraries require the user to pass a buffer. This API decision *could* end up forcing implementations to do more copying (from an internal buffer into a user buffer).

The other interesting side effect of forcing the caller to allocate/control the memory is that they may end up allocating memory before there's any data to read. That's of course that's all subject to how you expose concepts like sockets and websockets in deno but it's something to think about (https://eklitzke.org/goroutines-nonblocking-io-and-memory-usage)

File: Roadmap.md [hunk 5]
```diff
@@ -243,3 +243,151 @@ Fetch API:
...
+  // the error err. Doing so correctly handles I/O errors that happen after
+  // reading some bytes and also both of the allowed EOF behaviors.
+  //
+  // Implementations of Read are discouraged from returning a zero byte count
+  // with a nil error, except when p.byteLength == 0. Callers should treat a
+  // return of 0 and nil as indicating that nothing happened; in particular it
+  // does not indicate EOF.
+  //
+  // Implementations must not retain p.
+  async read(p: ArrayBufferView): Promise<ReadResult>;
```
Comment (davidfowl) [inline Roadmap.md]: Does there need to be a way to cancel the current read/write operation without signaling EOF?

File: Roadmap.md [hunk 6]
```diff
@@ -243,3 +243,151 @@ Fetch API:
...
+```ts
+// Copy copies from src to dst until either EOF is reached on src or an error
+// occurs. It returns the number of bytes copied and the first error encountered
+// while copying, if any.
+//
+// Because Copy is defined to read from src until EOF, it does not treat an EOF
+// from Read as an error to be reported.
+//
+// https://golang.org/pkg/io/#Copy
+async function copy(dst: Writer, src: Reader): Promise<number> {
```
Comment (davidfowl) [inline Roadmap.md]: This could be optimized if there was an interface that the reader could implement to write directly into the writer. If a reader is managing internal buffers then this logic might force copies. .NET has a virtual method `CopyToAsync` on the `Stream` abstraction that allows for this https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copytoasync?view=netcore-2.1

**Comment by benjamingr** [comment]
@davidfowl 

> +1, that's fundamental to making a good IO API. Most of the IO APIs in the various standard libraries require the user to pass a buffer. This API decision could end up forcing implementations to do more copying (from an internal buffer into a user buffer).

Pipelines look like a really interesting tradeoff for deno.

What's stopping an implementation from reading directly into the user supplied buffer rather than allocate the internal buffer? Looking at various I/O APIs it looks like that could be done reasonably which would let users control allocations. The two methods that have the least copies are:

 - The user allocates the buffer and the platform fills it.
 - The platform allocates the buffer and returns the user a view to it.

In both cases there is one copy but in the first one the user has more control over allocation. It's also possible to expose both (with `.read` and making Readers async iterable).

> That's of course that's all subject to how you expose concepts like sockets and websockets in deno but it's something to think about (https://eklitzke.org/goroutines-nonblocking-io-and-memory-usage)

That's a good point. Worth mentioning that TypeScript and Deno in this case are like C# (have explicit I/O yielding with async/await rather than implicit yielding) so the optimization discussed there will be possible with the proposed APIs.

File: Roadmap.md (hunk 5 above)
Comment (benjamingr) [reply Roadmap.md]: I think so, it's an interesting capability to expose with `shutdown(fd, SHUT_RD);`, worth mentioning neither Node nor Go support aborting a single read operation as far as I know and it hasn't been very commonly requested since these calls are typically cheap.

It's also worth mentioning we don't have proper cancellation semantics for async functions in JavaScript yet (CancelTokens were proposed but not really adopted and cancellable async functions are stuck). It might be another 2-3 years before we can have a good cancel story.

**Comment by davidfowl** [comment]
> What's stopping an implementation from reading directly into the user supplied buffer rather than allocate the internal buffer? 

Most of the time it's either to do with performance (reducing the number of sys calls for example). For example, the default FileStream in .NET uses a 4K buffer by default so if you pass in a buffer that's <= 4K, it'll copy into the user provided buffer.

It's the same for the `DeflateStream` in .NET (https://github.com/dotnet/corefx/blob/928873f0027f3c4fbea94c501990a7037ad27066/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L16). There's a default 8K buffer because that was determined to be an optimal chunk size for compression/decompression using zlib.

Some APIs use buffer pools (like webservers) to efficiently manage the memory for underlying IO operations and when they expose the Reader/Writer further up stack (like exposing the http request body), there's no way to expose those already pooled buffers. 

Copying isn't the end of the world, it's just a reality when you force the caller to allocate. It's easier to avoid that copying when the Reader/Writer interface directly with the OS, but it becomes more challenging in the higher layers of the system (when libraries expose their own Reader/Writer APIs). Everybody *could* end up storing their own internal buffer for optimizations so you end up with call chain like this:

The goal here is to have the user provide a buffer to read a simple websocket data frame. We have 3 readers:
1. The WebSocket Reader
2. The http request body Reader
3. The Socket Reader

The websocket API is built on top of the raw http request body Reader which itself is built on the socket reader. Every call to ReadAsync (it's C#) looks like this:

```
socket.ReceiveAsync(buffer) 4K
   httpRequest.Body.ReadAsync(buffer) 4K
      WebSocket.ReceiveAsync(buffer) 4K 
         Use(buffer)
```

> The platform allocates the buffer and returns the user a view to it.

This is the approach pipelines took (and it more mirrors what node did than go). Buffers are never allocated by the user by default. The underlying system will give you buffers and you need to tell us when you're done with them (there's an explicit call to say I'm done). Inherently, the pattern lets you choose if you want to make the copy or use the buffers provided by the implementation.

**Comment by benjamingr** [comment]
@davidfowl 

Thanks for that! If I understand correctly essentially the idea is that although it is possible to pass buffers directly to that OS that often breaks in practice since the APIs are layered and can't manage pooling well. 

> The websocket API is built on top of the raw http request body Reader which itself is built on the socket reader. Every call to ReadAsync looks like this:

That's a very good example. @robpike in your experience is the WebSocket example pattern described [above](https://github.com/ry/deno/pull/387#issuecomment-407678560) where each reader needs to make a copy common in practice? 

>  Inherently, the pattern lets you choose if you want to make the copy or use the buffers provided by the implementation.

That sounds very reasonable overall. Would you mind elaborating a little about how pipelines support both passing memory in and handling the buffering themselves? How do the use cases Ryan described above (like reading just the first 10 bytes) look?

I think _allowing_ the capability of passing memory in is super important. Not being able to pass the memory to read into bit me up a l lot of times before.

Also, given we expect `Reader` to be an interface people will implement - wouldn't supporting both of those be significantly harder? How are people expected to interact with `Pipelines` to write their own transforms?

**Comment by davidfowl** [comment]
> That sounds very reasonable overall. Would you mind elaborating a little about how pipelines support both passing memory in and handling the buffering themselves? How do the use cases Ryan described above (like reading just the first 10 bytes) look?

It doesn't support both, it always buffers so you *have* to copy if you want to get data into your own buffer. 

> How do the use cases Ryan described above (like reading just the first 10 bytes) look?

Today it's a loop that looks something like this:

```C#
public static async Task<ReadResult> ReadAsync(this PipeReader reader, int numBytes)
{
    while (true)
    {
        var result = await reader.ReadAsync();
        var buffer = result.Buffer;

        // EOF
        if (result.IsCompleted) return result;

        if (buffer.Length < numBytes)
        {
            reader.AdvanceTo(buffer.Start, buffer.End);
            continue;
        }

        // Return a result with the trimmed data.
        return new ReadResult(buffer.Slice(0, numBytes), result.IsCanceled, result.IsCompleted);
    }
}
```

And it could be used like this:

```C#
// Don't return until EOF or we're read the correct amount of data
await reader.ReadAsync(10);
```

Unlike the code in earlier examples, this actually doesn't yield the read until there are at least 10 bytes in the buffer so it simplifies the code users end up writing to consume data from the reader.

The exact same model is used for allowing users to pass in buffers at the cost of copying. But at least you *always* get a copy in that case, it isn't up to the implementation per se.

```C#

public async ValueTask<int> ReadAsync(this PipeReader reader, Memory<byte> destination)
{
    while (true)
    {
        var result = await reader.ReadAsync();
        var buffer = result.Buffer;
        try
        {
            if (!buffer.IsEmpty)
            {
                var count = Math.Min(buffer.Length, destination.Length);
                buffer = buffer.Slice(0, count);
                buffer.CopyTo(destination);
                return count;
            }

            if (result.IsCompleted)
            {
                return 0;
            }
        }
        finally
        {
            reader.AdvanceTo(buffer.End);
        }
    }
}
```

> I think allowing the capability of passing memory in is super important. Not being able to pass the memory to read into bit me up a l lot of times before.

Sure. I think the only concern is what I mentioned where multiple layers of buffering could be a concern. In the above model, passing in a buffer means you're opting into a copy.

> Also, given we expect Reader to be an interface people will implement - wouldn't supporting both of those be significantly harder? 

Yes, I'm not sure you want to support both on the interface. I think you need to pick a model and then adapt the other way around (You can potentially build one of top of the other).

> How are people expected to interact with Pipelines to write their own transforms?

There are 2 ways to do it:
- You use the default `Pipe` implementation which has a reader and writer. You write bytes into the writer and the reader can see them. The Pipe has lots of configurability (buffer sizes, scheduling, limits etc). It's similar to a channel in golang but specifically for bytes and optimized for IO.
- You can write your own PipeReader/PipeWriter that wraps another PipeReader/PipeWriter.

**Comment by benjamingr** [comment]
> Yes, I'm not sure you want to support both on the interface. I think you need to pick a model and then adapt the other way around (You can potentially build one of top of the other).

Thanks for clarifying and sorry for misunderstanding your previous comment about Pipelines and copies. That's pretty unfortunate but absolutely understandable.

@ry what about exposing base implementations rather than interfaces like the new .NET APIs do?

That is, expose a `Pipe` which is a `Reader/Writer` pair as well as Reader/Writer instances on the platform. If people don't need to implement the interface themselves (like promises for example) the platform can expose a `Symbol.asyncIterator` on them easily and the implementation burden is arguably easier.

Since you control the platform anyone writing a Reader/Writer is likely consuming another Reader/Writer anyway - this is like how one should almost never use the `Promise` constructor in Deno for instance since the API already returns promises.

File: Roadmap.md (hunk 5 above)
Comment (ry) [reply Roadmap.md]: @benjamingr shutdown is for half open TCP connections - not aborting read.

Comment (benjamingr) [reply Roadmap.md]: Sorry for the brainfart! That was incorrect from my end - I started writing about close semantics with abort and proceeded to write about cancellation semantics.

Comment (davidfowl) [reply Roadmap.md]: Right, it may not be common to cancel a single IO operation. So to abort the whole thing you need a ReaderCloser and you'd just call Close right? Usually you end up with a chain of calls where Reader/Writers decorate other Reader(s)/Writer(s) and you need to be able to abort/cancel that chain reliably.

File: Roadmap.md (hunk 6 above)
Comment (ry) [reply Roadmap.md]: It certainly would be nice to have [`sendfile(2)`](http://man7.org/linux/man-pages/man2/sendfile.2.html) support. 

I will leave adding a `copyTo` method for further revisions - as it's an optimization that won't exist in general - but thanks for bringing it up. (I'm not sure how/if Go does this...)

**Comment by ry** [comment]
@benjamingr wrote:
> Would you mind explaining why it's better that way (as an external method and not a symbol)?
> The biggest pro I can think of for making it external is that it might make implementing readers easier. The con is that if readers are expected to be worked with directly then this will be very verbose whenever a stream is iterated (which I assume is a very common if not the most common case). It is also different from what other languages did.
> Another concern is that if readers are not async-iterable people might pass around two types (readers and iterators).
> Other platforms solve that part with subclassing (which I don't really think we should do). It's also possible to expose a factory (like makeReader or something like that).

As you said, to make implementing readers easier. Don't exclude the possibility of having our core classes (like sockets and files) be both readers and async iterators - I just don't want every reader to be forced into implementing iterator semantics.

File: Roadmap.md [hunk 7]
```diff
@@ -243,3 +243,162 @@ Fetch API:
...
+  // returns n < p.byteLength, it may use all of p as scratch space during the
+  // call. If some data is available but not p.byteLength bytes, Read
+  // conventionally returns what is available instead of waiting for more.
+  //
+  // When Read encounters an error or end-of-file condition after successfully
+  // reading n > 0 bytes, it returns the number of bytes read. It may return the
+  // (non-nil) error from the same call or return the error (and n == 0) from a
+  // subsequent call. An instance of this general case is that a Reader
+  // returning a non-zero number of bytes at the end of the input stream may
+  // return either err == EOF or err == nil. The next Read should return 0, EOF.
```
Comment (benjamingr) [inline Roadmap.md]: Should this be updated based on the comment above about EOF not being an error on ReadResult and `Read` -> `read`?

File: Roadmap.md [hunk 8]
```diff
@@ -243,3 +243,162 @@ Fetch API:
...
+  //
+  // Callers should always process the n > 0 bytes returned before considering
+  // the error err. Doing so correctly handles I/O errors that happen after
+  // reading some bytes and also both of the allowed EOF behaviors.
+  //
+  // Implementations of Read are discouraged from returning a zero byte count
```
Comment (benjamingr) [inline Roadmap.md]: s/Read/read/

File: Roadmap.md (hunk 3 above)
Comment (benjamingr) [reply Roadmap.md]: I understood it to also imply that implementations should not store references to the TypedArrays passed to it as well (rather than just copies). I wonder if there is a better way to say this.

File: Roadmap.md (hunk 5 above)
Comment (benjamingr) [reply Roadmap.md]: @davidfowl You'd (hopefully) `.close` one which would close the one it's wrapping and so on cascading to close the calls and reject all the pending operations.

Ideally I'd want it to abort (and not reject) the operations but since we don't have promise cancellations or cancellation semantics yet - rejecting pending `Read`s on `Close` is probably the best we may be able to do but it's worth prototyping and checking for footguns.

**Comment by robpike** [comment]
@ benjamingr I regret nothing about the current I/O interfaces in Go. They are clean and simple. The subtleties are all around corner cases not easily capture by the type system, such as what the caller can assume about, and do with, the buffer passed to Read or Write.

Because of the language and platform, you need to face non-blocking I/O. That could have a big effect on your design, but having spent my career trying to prevent the need for non-blocking I/O, I am not your best guide here.

File: Roadmap.md (hunk 6 above)
Comment (davidfowl) [reply Roadmap.md]: Here it is https://github.com/golang/go/blob/15ac56fd60a28879fd65f6d761cc30c4c57ec0f1/src/io/io.go#L181.

> // WriterTo is the interface that wraps the WriteTo method.
> //
> // WriteTo writes data to w until there's no more data to write or
> // when an error occurs. The return value n is the number of bytes
> // written. Any error encountered during the write is also returned.
> //
> // The Copy function uses WriterTo if available.

**Comment by benjamingr** [comment]
Just to clarify (in case it wasn't clear) - after the discussion (thank you) and the addition of `readerIterator` for compatibility with new ES features I am fine with the interface.

(I also think David raises a very interesting point but am still for letting users avoid allocations by passing buffers)

**Comment by davidfowl** [comment]
> (I also think David raises a very interesting point but am still for letting users avoid allocations by passing buffers)

The ironic thing is that users really can't always avoid allocations when you stack these things when you force them to allocate.

File: Roadmap.md (hunk 8 above)
Comment (ry) [reply Roadmap.md]: Done

File: Roadmap.md (hunk 7 above)
Comment (ry) [reply Roadmap.md]: Done

**Comment by ry** [comment]
@piscisaureus PTAL - I'd like to land this without many more changes and fix up in later commits - as this has been hanging around for too long now.
